### PR TITLE
Enable attachment of binary files as charm resources

### LINKED ISSUE #1000: Unable to upload .tar file as local resource to deployed charmed application
## The problem

When trying to run the integration tests for the slurmd charmed operator, our CI pipeline fails with the following stacktrace:

```python
Traceback (most recent call last):
  File "/home/runner/work/slurmd-operator/slurmd-operator/tests/integration/test_charm.py", line 48, in test_build_and_deploy
    await asyncio.gather(
  File "/home/runner/work/slurmd-operator/slurmd-operator/.tox/integration/lib/python3.10/site-packages/juju/model.py", line 1820, in deploy
    resources = await self.add_local_resources(application_name,
  File "/home/runner/work/slurmd-operator/slurmd-operator/.tox/integration/lib/python3.10/site-packages/juju/model.py", line 2052, in add_local_resources
    data = p.read_text() if p.exists() else ''
  File "/usr/lib/python3.10/pathlib.py", line 1135, in read_text
    return f.read()
  File "/usr/lib/python3.10/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte
```

This error is raised when we try to attach our resource  `lbnl-nhc-1.4.3.tar.gz` to the slurmd application. NHC (Node Health Check) is a utility for measuring the health of compute nodes in your cluster; it is commonly used to collect observability metrics. When we try to attach NHC, `pathlib.Path(...).read_text(...)` tries to decode the .tar.gz file as an utf-8 text file rather than a binary file. This causes python-libjuju to raise the above `UnicodeDecodeError` which then torpedoes our CI pipeline. The block of code below is what is responsible for this issue: 

https://github.com/juju/python-libjuju/blob/2581b0ced1df6201c6b7fd8cc0b20dcfa9d97c51/juju/model.py#L2050-L2052

This could potentially be fixed by switch the implementation to `pathlib.Path(...).read_bytes(...)`

## Steps to reproduce

1. _Assume that you already have a cloud and Juju controller set up..._
2. `git clone -b jaime/drop-focal-support git@github.com:jaimesouza/slurmd-operator.git`
3. `cd slurmd-operator`
4. `tox run -e integration`
5. Wait a few minutes...
6. See stacktrace with `UnicodeDecodeError`

### PR DESCRIPTION
**Pull request by nrobinaubertin**
#### Description

This enables the attachment of binary files as charm resources.

Fixes: https://github.com/juju/python-libjuju/issues/1000

### CODE REVIEW

File: juju/model.py [hunk 1]
```diff
@@ -2078,7 +2078,7 @@ def _upload(self, data, path, app_name, res_name, res_type, pending_id):
 
         headers['Content-Type'] = 'application/octet-stream'
         headers['Content-Length'] = len(data)
-        headers['Content-Sha384'] = hashlib.sha384(bytes(data, 'utf-8')).hexdigest()
+        headers['Content-Sha384'] = hashlib.sha384(data).hexdigest()
```
Comment (dimaqq) [inline juju/model.py]: This effectively changes the argument type from `str` to `bytes`.

There's one more call site at line 2141 that would have to be updated to match.

P.S. please add an explicit type hint on the argument.

**Comment by dimaqq** [comment]
@nrobinaubertin this PR needs some kind of test.

ideally an integration test, or 
a series of steps (this exact charm, this bootstrap, this deploy with this binary artifact)

**Review by gfouillet — CHANGES_REQUESTED** [review]
Nothing to more than dima.

* Please add a test that would trigger the bug if not fixed
* take care of all call site of `_upload` to ensure `data` type consistency.

**Comment by dimaqq** [comment]
Niels, I see that the PR where you needed this was closed.

Do you have the bandwidth to update this PR?

If not, I can maybe create an issue based on the PR and eventually someone else can pick this up. What you're fixing is basically a bug, as binary resources should be supported.

**Comment by dimaqq** [comment]
This PR needs to be rebased since #1186 has been merged.

**Comment by nrobinaubertin** [comment]
> I see that the PR where you needed this was closed.

I'm still using the feature but I'm using my patch locally

> Do you have the bandwidth to update this PR?

Not right now, so if you feel we should close this PR, we can do it and I may reopen it later.  
The patch in itself is tiny and straightforward, but I would need a bit of spare time to write the tests

> If not, I can maybe create an issue based on the PR and eventually someone else can pick this up. What you're fixing is basically a bug, as binary resources should be supported.

There is already an issue, #1000

**Comment by dimaqq** [comment]
- rebased
- fixed the yaml code path
- added type hints

File: juju/model.py (hunk 1 above)
Comment (dimaqq) [reply juju/model.py]: done now

**Comment by dimaqq** [comment]
Moved to #1218

**Pull request closed without merging**